Runtime: vim.inspect: Show file+line for functions

runiq opened the pull request:
Makes `vim.inspect()` more useful for functions by showing where they are defined.

Since `vim.inspect` shouldn't be used in any hot code paths anyways, I hope using the `debug` module (which is said to have non-optimal performance) in this way is appropriate.

Side note: This is not terribly for functions defined in C (like, for example, API functions).

Example output of `print(vim.inspect(vim.lsp.handlers))`:

```lua
{
  ["$/progress"] = <function 1: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["callHierarchy/incomingCalls"] = <function 2: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["callHierarchy/outgoingCalls"] = <function 3: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/codeAction"] = <function 4: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/completion"] = <function 5: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/declaration"] = <function 6: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/definition"] = <function 7: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/documentHighlight"] = <function 8: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/documentSymbol"] = <function 9: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/formatting"] = <function 10: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/hover"] = <function 11: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/implementation"] = <function 12: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/publishDiagnostics"] = <function 13: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp.lua:1250>,
  ["textDocument/rangeFormatting"] = <function 14: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/references"] = <function 15: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/rename"] = <function 16: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/signatureHelp"] = <function 17: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["textDocument/typeDefinition"] = <function 18: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["window/logMessage"] = <function 19: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["window/showMessage"] = <function 20: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["window/showMessageRequest"] = <function 21: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["window/workDoneProgress/create"] = <function 22: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["workspace/applyEdit"] = <function 23: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["workspace/configuration"] = <function 24: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["workspace/executeCommand"] = <function 25: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>,
  ["workspace/symbol"] = <function 26: /home/ajaam/.local/opt/nvim/share/nvim/runtime/lua/vim/lsp/handlers.lua:393>
}
```

Example output of `print vim.inspect(vim.api.nvim_command))`:

```lua
<function 1: [C]>
```

runiq committed:
Runtime: vim.inspect: Show file+line for functions

Makes vim.inspect more useful for functions by showing where they are
defined.

bfredl commented on runtime/lua/vim/inspect.lua:
```diff
@@ -288,6 +288,16 @@ function Inspector:putTable(t)
+  self:puts(
+    '<',
+    'function ', self:getId(f), ': ',
+    info.source, ':', info.linedefined,
```
maybe we should omit `':', info.linedefined` when it is -1 ?

runiq replied:
Sure, will do.

runiq committed:
Request only source info ('S')

runiq committed:
Streamline function source location display

- Remove `@` prefix for files
- Remove `=` prefix for C functions
- Remove line count if < 0

runiq commented:
As you can see, this isn't terribly useful in more cases than I'd like to admit:

- All functions in `print(vim.inspect(vim.lsp.handlers))` point to the same line as they are assigned in a loop.
- All you can tell about C functions (e.g. API functions) is that they are C functions. No source code locations. (Incidentally, all API functions are C functions.)

That being said, it's still an improvement over the status quo IMHO, especially for debugging your own Lua functions.

justinmk commented:
I suppose this is fine, but can the path be shortened? At least home should be `~/`, and maybe also relative to CWD when appropriate.
